Cython/Utility/ObjectHandling.c

```diff
@@ -2000,16 +2000,83 @@ static int __Pyx_TryUnpackUnboundCMethod(__Pyx_CachedCFunction* target) {
     target->method = method;
 #if CYTHON_COMPILING_IN_CPYTHON
     #if PY_MAJOR_VERSION >= 3
-    // method descriptor type isn't exported in Py2.x, cannot easily check the type there
     if (likely(__Pyx_TypeCheck(method, &PyMethodDescr_Type)))
+    #else
+    // method descriptor type isn't exported in Py2.x, cannot easily check the type there.
+    // Therefore, reverse the check to the most likely alternative
+    // (which is returned for class methods)
+    if (likely(!PyCFunction_Check(method)))
     #endif
     {
         PyMethodDescrObject *descr = (PyMethodDescrObject*) method;
         target->func = descr->d_method->ml_meth;
         target->flag = descr->d_method->ml_flags & ~(METH_CLASS | METH_STATIC | METH_COEXIST | METH_STACKLESS);
-    }
+    } else
+#endif
+    // bound classmethods need special treatment
+#if defined(CYTHON_COMPILING_IN_PYPY)
+    // In these functions are regular methods, so just do
+    // the self check
+#elif PY_VERSION_HEX >= 0x03090000
+    if (PyCFunction_CheckExact(method))
+#else
+    if (PyCFunction_Check(method))
+#endif
+    {
+        PyObject *self = NULL;
+        int self_found = 0;
+#if CYTHON_COMPILING_IN_LIMITED_API || CYTHON_COMPILING_IN_PYPY
+        self = PyObject_GetAttrString(method, "__self__");
+        if (!self) {
+            PyErr_Clear();
+        }
+#else
+        self = PyCFunction_GET_SELF(method);
+#endif
+        self_found = (self && self != Py_None);
+#if CYTHON_COMPILING_IN_LIMITED_API
+        Py_XDECREF(self);
+#endif
+        if (self_found) {
+            // This is almost certainly a pessimization - we have a bound classmethod
+            // which will be passed "self". We therefore just create a lambda function
+            // to ignore the "self" argument. However, there are only 5 builtin
```

Comment:
Would a `staticmethod` work here as a wrapper? That's probably simpler and faster than going through the whole Python compiler.

Reply:
I don't think so (or I can't make it work) - I don't think a `staticmethod` ignores the first argument - I think most of what it does is in `__get__` and we're now too late to benefit from that.

I don't really like what I've currently done though so happy with better alternatives

Reply:
Something like this should take care of the problem
```c
// pun intended
PyObject *SelflessCall(PyObject *method, PyObject *args, PyObject *kwargs) {
    // selfless_args = ...
    return PyObject_Call(method, selfless_args, kwargs);
}
```

```c
if (self_found) {
    PyMethodDef def = {
        .ml_name = ...,
        .ml_meth = (PyCFunction)SelflessCall,
        .ml_flags = METH_VARARGS | METH_KEYWORDS,
        .ml_doc = NULL
    };
    target->method = PyCFunction_New(def, method);
}
```

You can actually pass anything as a second argument for `PyCFunction_New`, and it will appear as the first argument in `SelflessCall`, AFAIK

Reply:
The `PyMethodDef` needs to be static I think. Although that shouldn't be an insurmountable problem. I'll look into that suggestion...

Reply:
Correction, we probably need to pymem_malloc PyMethodDef  and at best have it accounted in custom pyobject alongside method, as PyCFunctionObject does not free PyMethodDef's memory on destruction

```c
static void
meth_dealloc(PyCFunctionObject *m)
{
    // The Py_TRASHCAN mechanism requires that we be able to
    // call PyObject_GC_UnTrack twice on an object.
    PyObject_GC_UnTrack(m);
    Py_TRASHCAN_BEGIN(m, meth_dealloc);
    if (m->m_weakreflist != NULL) {
        PyObject_ClearWeakRefs((PyObject*) m);
    }
    // Dereference class before m_self: PyCFunction_GET_CLASS accesses
    // PyMethodDef m_ml, which could be kept alive by m_self
    Py_XDECREF(PyCFunction_GET_CLASS(m));
    Py_XDECREF(m->m_self);
    Py_XDECREF(m->m_module);
    PyObject_GC_Del(m);
    Py_TRASHCAN_END;
}
```

Reply:
Or just a tuple, something like

```c
...
PyTupleObject *tup = PyTuple_New(2);
PyTuple_SetItem(tup, 0, method);
PyTuple_SetItem(tup, 1, PyCapsule_New((void*)def, NULL, struct_capsule_destructor));

target->method = PyCFunction_New(&def, tup);
```

Reply:
Alright, I folded this conversation a little, @da-woods are you ok with the option that I will create anothe PR based on your branch implementing what I've described above and we can merge both then?

Reply:
> The `PyMethodDef` needs to be static I think. Although that shouldn't be an insurmountable problem. I'll look into that suggestion...

If the name will always be the same, then yeah, we can make it static global struct
Should it be the same name though, I think about this function more like a wrapper, so it's pribably nice to preserve the name of the original funciton

Reply:
> If the name will always be the same

I don't think it should be possible to access the name from anywhere external do I'd go with the static version as being simpler.

> I will create anothe PR based on your branch implementing what I've described above

You can do if you want and have time.

Reply:
> I don't think it should be possible to access the name from anywhere external do I'd go with the static version as being simpler.

Yeah, I've looked into that and it seems to be the case

> You can do if you want and have time.

I'd like to give it a try I've found this to be an interesting thing to implement

Reply:
Having static PyMethodDef is surely nicer than having dynamic Py_Capsule